internal/xds/rbac/rbac_engine.go

```diff
@@ -48,6 +53,95 @@ func NewEngine(policy *v3rbacpb.RBAC) (*Engine, error) {
 	return &Engine{policies: policies}, nil
 }
 
+// Helper functions here which put connection in and out of the context.
+// This is used for authorization interceptors, so do we put it here or in authorization package?
+type connectionKey struct{}
+
+func getConnection(ctx context.Context) net.Conn {
+	conn, _ := ctx.Value(connectionKey{}).(net.Conn)
+	return conn
+}
+
+// SetConnection adds the connection to the context to be able to get
+// information about the destination ip and port for an incoming RPC.
+func SetConnection(ctx context.Context, conn net.Conn) context.Context {
+	return context.WithValue(ctx, connectionKey{}, conn)
+}
+
+// NewRPCData takes a incoming context (should be a context representing state
+// needed for server RPC Call with headers and connection piped into it) and the
+// method name of the Service being called server side and populates an RPCData
+// struct ready to be passed to the RBAC Engine to find a matching policy.
+func NewRPCData(ctx context.Context, fullMethod string) (*RPCData, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("error retrieving metadata from incoming ctx")
+	}
+
+	pi, ok := peer.FromContext(ctx)
+	if !ok {
+		return nil, errors.New("error retrieving peer info from incoming ctx")
+	}
+
+	// You need the connection in order to find the destination address and port
+	// of the incoming RPC Call.
+	conn := getConnection(ctx)
+	_, dPort, err := net.SplitHostPort(conn.LocalAddr().String())
+	if err != nil {
+		return nil, err
+	}
+	dp, err := strconv.ParseUint(dPort, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsState := pi.AuthInfo.(credentials.TLSInfo).State // TODO: Handle errors on type conversion?
+	pName, err := findPrincipalNameFromCerts(tlsState.PeerCertificates)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RPCData{
+		MD:              md,
+		PeerInfo:        pi,
+		FullMethod:      fullMethod,
+		DestinationPort: uint32(dp),
+		DestinationAddr: conn.LocalAddr(),
+		PrincipalName:   pName,
+	}, nil
+}
+
+// findPrincipalNameFromCerts extracts the principal name from the TLS
+// Certificates according to the RBAC configuration logic - "The URI SAN or DNS
+// SAN in that order is used from the certificate, otherwise the subject field
+// is used."
+func findPrincipalNameFromCerts(certs []*x509.Certificate) (string, error) {
+	if len(certs) == 0 {
+		return "", errors.New("there are no certificates present")
+	}
+	// Loop through and find URI SAN if present.
+	for _, cert := range certs {
+		for _, uriSAN := range cert.URIs {
+			return uriSAN.String(), nil
```

Comment:
Thanks Zach for working on this. 

I feel the principal name is not handled properly.
A certificate could have multiple URI SANs and DNS SANs. Shouldn't we be returning all the values here?
Also if you look into the Authenticated Matcher, it would compare the value in policy with the URI SANs, and if no match were found, then it would compare with DNS SANs, and finally subject field.
https://github.com/envoyproxy/envoy/blob/0fae6970ddaf93f024908ba304bbd2b34e997a51/source/extensions/filters/common/rbac/matchers.cc#L174
So we may need different fields for uri, dns and subject in RPCData.